### PR TITLE
Add upsample_cfg support in FPN

### PR DESCRIPTION
**Pull request by Johnson-Wang**
Small changes. 
1. Added upsample config support in FPN. 
2. Support three sources of the extra convs: inputs, laterals and outs. These three connections all exist in different codebases. This is to support model conversion to mmdet.

### CODE REVIEW

File: mmdet/models/necks/fpn.py [hunk 1]
```diff
@@ -34,6 +35,14 @@ class FPN(nn.Module):
         norm_cfg (dict): Config dict for normalization layer. Default: None.
         act_cfg (str): Config dict for activation layer in ConvModule.
             Default: None.
+        upsample_cfg (dict): Config dict for interpolate layer.
+            Default: `dict(mode='nearest')`
+        extra_convs_source (str, optional): The source feature map of the extra
```
Comment (Johnson-Wang) [inline mmdet/models/necks/fpn.py:40]: `extra_convs_source` have some overlaps with `extra_convs_on_inputs`. But to keep compatible with previous configs, I have left the key `extra_convs_on_inputs` intact.

**Comment by ZwwWayne** [comment]
Need unit tests to to test the API `upsample_cfg` and `extra_convs_source`.

File: tests/test_necks.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            feats[i].shape[2:] == outs[i].shape[2:]
+
+    # Tests for fpn with no extra convs (pooling is used instead)
+    fpn_model = FPN(
+        in_channels=in_channels,
+        out_channels=out_channels,
+        start_level=1,
+        add_extra_convs=False,
+        num_outs=5)
+    fpn_model(feats)
```
Comment (ZwwWayne) [inline tests/test_necks.py:63]: Maybe we need to further check the size of the output?

File: mmdet/models/necks/fpn.py (hunk 1 above)
Comment (hellock) [reply mmdet/models/necks/fpn.py:40]: We can deprecate this option and remove it in future releases.

File: mmdet/models/necks/fpn.py [hunk 3]
```diff
@@ -34,6 +35,14 @@ class FPN(nn.Module):
...
+            convs. `None` means it is specified by `extra_convs_on_inputs`.
+            Otherwise, only the following modes are allowed
+             'inputs': Last feature map of neck inputs (i.e. backbone feature).
+             'laterals':  Last feature map after lateral convs
+             'outputs': The last output feature map after fpn convs
```
Comment (hellock) [inline mmdet/models/necks/fpn.py:45]: I suggest using `input`, `lateral`, `output`.

File: mmdet/models/necks/fpn.py [hunk 4]
```diff
@@ -63,7 +72,9 @@ def __init__(self,
                  no_norm_on_lateral=False,
                  conv_cfg=None,
                  norm_cfg=None,
-                 act_cfg=None):
+                 act_cfg=None,
+                 upsample_cfg=dict(mode='nearest'),
+                 extra_convs_source=None):
```
Comment (hellock) [inline mmdet/models/necks/fpn.py:77]: This argument may be placed after `add_extra_convs`.

File: mmdet/models/necks/fpn.py (hunk 1 above)
Comment (hellock) [reply mmdet/models/necks/fpn.py:40]: We may use a single argument `add_extra_convs` to specific behaviors.
- False
- `'on_input'`
- `'on_lateral'`
- `'on_output'`

File: mmdet/models/necks/fpn.py (hunk 3 above)
Comment (Johnson-Wang) [reply mmdet/models/necks/fpn.py]: Removed extra_convs_source according to the last comment.

File: mmdet/models/necks/fpn.py [hunk 5]
```diff
@@ -22,10 +22,18 @@ class FPN(nn.Module):
...
-        extra_convs_on_inputs (bool): Whether to apply extra conv on
-            the original feature from the backbone. Default: False.
+        add_extra_convs (bool | str): If bool, it decides whether to add conv
+            layers on top of the original feature maps. Default to False.
+            If True, its actual mode is specified by `extra_convs_on_inputs`.
+            If str, it specifies the source feature map of the extra convs.
+            Only the following options are allowed
+             'on_input': Last feat map of neck inputs (i.e. backbone feature).
+             'on_lateral':  Last feature map after lateral convs.
+             'on_output': The last output feature map after fpn convs.
```
Comment (hellock) [inline mmdet/models/necks/fpn.py:32]: The format will not be correctly parsed in the doc.
```suggestion
            Only the following options are allowed
            
            - 'on_input': Last feat map of neck inputs (i.e. backbone feature).
            - 'on_lateral':  Last feature map after lateral convs.
            - 'on_output': The last output feature map after fpn convs.
```

**Comment by hellock** [comment]
The module `necks` is missing in `api.rst`.

File: docs/api.rst [hunk 6]
```diff
@@ -80,6 +80,11 @@ dense_heads
 .. automodule:: mmdet.models.dense_heads
     :members:
 
+necks
+^^^^^^^^^^^^
+.. automodule:: mmdet.models.necks
+    :members:
```
Comment (hellock) [inline docs/api.rst:86]: It may be moved before `dense_heads` following the conventional order.

File: mmdet/models/necks/fpn.py [hunk 7]
```diff
@@ -22,10 +22,21 @@ class FPN(nn.Module):
...
             build the feature pyramid. Default: -1, which means the last level.
-        add_extra_convs (bool): Whether to add conv layers on top of the
-            original feature maps. Default: False.
-        extra_convs_on_inputs (bool): Whether to apply extra conv on
-            the original feature from the backbone. Default: False.
+        add_extra_convs (bool | str): If bool, it decides whether to add conv
+            layers on top of the original feature maps. Default to False.
+            If True, its actual mode is specified by `extra_convs_on_inputs`.
+            If str, it specifies the source feature map of the extra convs.
+            Only the following options are allowed
```
Comment (hellock) [inline mmdet/models/necks/fpn.py:29]: The docstring should follow the rst syntax. An empty line is needed before list, but not needed between items of the list.

File: mmdet/models/necks/fpn.py [hunk 8]
```diff
@@ -22,10 +22,19 @@ class FPN(nn.Module):
...
+
+            - 'on_input': Last feat map of neck inputs (i.e. backbone feature).
+            - 'on_lateral':  Last feature map after lateral convs.
+            - 'on_output': The last output feature map after fpn convs.
+        extra_convs_on_inputs (bool, deprecated): Whether to apply extra convs
```
Comment (hellock) [inline mmdet/models/necks/fpn.py:34]: Since this argument is deprecated, we can update existing configs that are using it.

Comment (Johnson-Wang) [reply mmdet/models/necks/fpn.py:34]: Ok, by the way, I just want to make sure that even though this argument is deprecated, we should not delete it directly. Only changing the configs would suffice. Am I correct?

Comment (hellock) [reply mmdet/models/necks/fpn.py:34]: Yes. We would like to switch to the new argument in later versions, but also provide backward compatibility for existing forks and projects.

**Pull request merged**